docs/design/002-containerized-gadgets.md

```diff
@@ -0,0 +1,602 @@
+
+NOTE: It's under development at the time of writing this document in
+https://github.com/inspektor-gadget/inspektor-gadget/pull/1825.
+
+#### `socket_filter` programs
+
+Attach ebpf programs of type `BPF_PROG_TYPE_SOCKET_FILTER`. TODO: What's special about these?
+
+#### Socket enricher
+
+Use the socket enricher from the ebpf code. This can be done in a very similar way to the
+`gadget_mntns_filter_map`: Inspektor Gadget looks for this map, automatically replaces and keeps it
+updated with the right one.
+
+#### Custom parameters
+
+Some gadgets define constants on the eBPF code to change its behavior. Those constants should be
+exposed to the user as parameters.
+
+It could be implemneted by using the pkg/params package.
+
+
+#### Enum convert
+
+Convert an enum to a string. Examples: pktTypeNames, qTypeNames, rCodeNames, etc. Some could use BTF
+from the kernel, but not all.
+
+Mauricio: I think we can use the BTF from the program too.
+
+#### String edit
+
+string manipulation routines. Example: parseLabelSequence to replace dns strings to dotted names.
+
+#### Bitfield
+
+Convert bitfield from an event to a human readable string. Example: bind's [optionsToString](https://github.com/inspektor-gadget/inspektor-gadget/blob/b57f2bae31a46b40d8e0204b85099ae37f15d21d/pkg/gadgets/trace/bind/tracer/tracer.go#L154).
+
+#### Iface index
+
+Convert network interface index to string. Example: bind's [BoundDevIf](https://github.com/inspektor-gadget/inspektor-gadget/blob/b57f2bae31a46b40d8e0204b85099ae37f15d21d/pkg/gadgets/trace/bind/tracer/tracer.go#L236-L250).
+
+#### Custom user space logic
+
+There are some gadgets that have a lot of logic in user space, to support this gadgets we'll need to
+run custom code they provide. This presents big challenges (security, flexibility) and it's out of
+scope for now. We can revisit this later on if a needed arises.
+
+### Gadgets
+
+This is the list of the existing gadgets and the considerations to port them to containerized
+gadgets. The purpose of this section is to describe the needed features and not to provide a
+solution for them.
+
+#### advise network-policy
+
+This one has a lot of custom logic in user space, it requires [custom user space
+logic](#custom-user-space-logic)
+
+#### advise seccomp-profile
+
+TODO
+
+#### audit seccomp
+
+TODO
+
+#### profile block-io
+
+TODO
+
+#### profile cpu
+
+TODO
+
+#### profile tcprtt
+
+TODO
+
+#### snapshot process
+
+It supports a compatibility mode to get processes when iterators are not available ->
+https://github.com/inspektor-gadget/inspektor-gadget/blob/e91a93bb2c0a7d05ed6ac2689b89ec7465923dcf/pkg/gadgets/snapshot/process/tracer/tracer.go#L216.
+We'll lose that in the containerized version.
+
+#### snapshot socket
+
+Inspektor Gadget should automatically run the iter/ programs on the different network namespaces of
+the containers. It's still TBD how a gadget will define that it wants this behavior.
```

Comment:
It depends on the type of iterator:
* iter/task, iter/task_file: relative to the pid namespace. But since pid namespaces are hierarchical, to be executed on the host pid namespace. If running with hostPID=true, easy to do. If running without hostPID, it is done with a host systemd unit thanks to pkg/utils/bpf-iter-ns/
* iter/tcp, iter/udp: relative to the net namespace. To be run in all network namespaces of interest (containers selected with the usual filter flags like `--container`)
* iter/bpf_map_elem: relative to a map. Unsupported.

Reply:
Thanks, added this comment in the "BPF Iterators (a.k.a snapshotters)" section.